data/tutorials/language/0lg_09_env.md

```diff
@@ -0,0 +1,562 @@
+1. Assuming it was possible, such an algorithm would declare that quicksort and bubble sort are equal. That would mean those procedures are substitutable, which is not the case.
+
+It may seem counterintuitive that classes of objects of the same kind (i.e., having the same type) exist where equality between objects does not make sense. High school mathematics does not provide examples of those classes. But in the case of computing procedures seen as functions, equality isn't the right tool to compare them.
+
+Pattern matching does not allow to inspect a function. Catch-all patterns can match against a function, but it is useless.
+```ocaml
+# match Fun.id with id -> ();;
+- : unit = ()
+```
+
+### Calling Functions
+
+In OCaml, no symbol is needed to express passing an argument to a function. When several expressions are written side by side, the leftmost one should be a function. All the others are parameters. In OCaml, parentheses serve a single purpose: associating expressions to create subexpressions.
+```ocaml
+# max (21 * 2) (int_of_string "713");;
+- : int = 713
+```
+
+There are two alternative ways to call functions: using the application `@@` operator or the pipe `|>` operator.
+```ocaml
+# ( @@ );;
+- : ('a -> 'b) -> 'a -> 'b = <fun>
+
+# sqrt 9.0;;
+- : float = 3.
+
+# sqrt @@ 9.0;;
+- : float = 3.
+```
+
+The application operator applies a parameter (on the right) to a function (on the left). It is useful when chaining several calls, as it avoids writing parentheses. Here is what it can look like:
+```ocaml
+# truncate (sqrt (float_of_int (int_of_string "81")));;
+- : int = 9
+
+# truncate @@ sqrt @@ float_of_int @@ int_of_string "81";;
+- : int = 9
+```
+
+The pipe operator also allows avoiding parentheses but in reversed order:
+```ocaml
+# ( |> );;
+- : 'a -> ('a -> 'b) -> 'b = <fun>
+
+# "81" |> int_of_string |> float_of_int |> sqrt |> truncate;;
+- : int = 9
+```
+This is just like a Unix shell pipe.
+
+### Anonymous Functions
+
+As citizens of the same level as other values, functions don't have to be bound to a name to exist (although some must, but this will be explained later). Function values not bound to names are called _anonymous functions_. Here are a couple of examples:
+```ocaml
+# fun x -> x;;
+- : 'a -> 'a = <fun>
+
+# fun x -> x * x;;
+- : int -> int = <fun>
+
+# fun s t -> s ^ " " ^ t ;;
+- : string -> string-> string = <fun>
+
+# function [] -> None | x :: _ -> Some x;;
+- : 'a list -> 'a option = <fun>
+```
+
+In order, here is what they do:
+- The identity function that takes anything, returns it unchanged
+- The square function takes an integer, returns its square
+- A function taking two strings, returning their concatenation, with a space character in between
+- A function taking a list, returning `None` if the list is empty, and the tail of the list otherwise
```

Comment:
```suggestion
- A function taking a list returns `None` if the list is empty, and the tail of the list otherwise
```
... returns the tail of the list otherwise?